src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleasePermissionService.cs

```diff
@@ -0,0 +1,87 @@
+        private readonly IUserService _userService;
+
+        public ReleasePermissionService(
+            IPersistenceHelper<ContentDbContext> persistenceHelper,
+            ContentDbContext contentDbContext,
+            IPublicationRepository publicationRepository,
+            IUserReleaseRoleRepository userReleaseRoleRepository,
+            IUserService userService)
+        {
+            _persistenceHelper = persistenceHelper;
+            _contentDbContext = contentDbContext;
+            _publicationRepository = publicationRepository;
+            _userReleaseRoleRepository = userReleaseRoleRepository;
+            _userService = userService;
+        }
+
+        public async Task<Either<ActionResult, List<ManageAccessPageContributorViewModel>>> GetManageAccessPageContributorList(Guid releaseId)
+        {
+            return await _persistenceHelper
+                .CheckEntityExists<Release>(releaseId,
+                    query =>
+                        query.Include(r => r.Publication))
+                .OnSuccessDo(release => _userService
+                    .CheckCanUpdateReleaseRole(release.Publication, ReleaseRole.Contributor))
+                .OnSuccess(async release =>
+                {
+                    var allLatestReleases = await _publicationRepository
+                        .GetLatestReleases(release.PublicationId);
+
+                    var allLatestReleaseIds = allLatestReleases
+                        .Select(r => r.Id)
+                        .ToList();
+
+                    var allContributorReleaseRoles = await _userReleaseRoleRepository.GetAllReleaseRoles(
+                        ReleaseRole.Contributor,
+                        allLatestReleaseIds.ToArray());
+
+                    var allPublicationContributors = allContributorReleaseRoles
+                        .Select(releaseRole => releaseRole.User)
+                        .Distinct()
+                        .ToList();
+
+                    return allPublicationContributors
+                        .Select(user =>
+                        {
+                            var roleForThisRelease = allContributorReleaseRoles.SingleOrDefault(releaseRole =>
+                                releaseRole.UserId == user.Id
+                                && releaseRole.ReleaseId == releaseId);
+
+                            return new ManageAccessPageContributorViewModel
+                            {
+                                UserId = user.Id,
+                                UserFullName = user.DisplayName,
+                                ReleaseId = releaseId,
+                                // Some don't have a contributor release role for this release, as they are a
+                                // contributor on a different release on the same publication
+                                ReleaseRoleId = roleForThisRelease?.Id,
+                            };
+                        }).ToList();
+                });
+        }
```

Comment:
I remember we talked about only fetching the data back needed for each release when it's expanded as working out the roles for each release when there's a lot of previous releases could be quite heavy. However I see here we effectively do all that work for the whole Publication every time this is called for each of the previous releases.

We know what all the releases are, we know all the contributing team users, we know all the roles. 

If this method was parameterised by publication id rather than release id then in one call with similar effort it could return a view model containing 

* Users (userId, displayName)
* Releases (releaseId, name, isLatest - ordered by time period)
* UserReleaseRoles (userId, releaseId, releaseRole)

and leave the FE to map them into the display. It would have all the data it needs for the whole page.

At the moment the same effort happens for each release call as they are expanded, to return a list of models for each user.

I suppose the list of UserReleaseRoles could grow quite big and is it's unnecessary to return all of them if they are only interested in managing the latest Release, but I'm just pointing out the same information is being retrieved from the db every time a Release is expanded so might be worth chatting about it.